src/pip/_internal/utils/unpacking.py

```diff
@@ -0,0 +1,286 @@
+# mypy: strict-optional=False
+
+from __future__ import absolute_import
+
+import logging
+import mimetypes
+import os
+import shutil
+import stat
+import tarfile
+import zipfile
+
+from pip._internal.exceptions import InstallationError
+from pip._internal.utils.misc import (
+    current_umask,
+    ensure_dir,
+    file_contents,
+    has_leading_dir,
+    is_svn_page,
+    rmtree,
+    split_leading_dir,
+)
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import Optional
+    from pip._internal.models.link import Link
+    from pip._internal.utils.hashes import Hashes
+
+logger = logging.getLogger(__name__)
+
+WHEEL_EXTENSION = '.whl'
+BZ2_EXTENSIONS = ('.tar.bz2', '.tbz')
+XZ_EXTENSIONS = ('.tar.xz', '.txz', '.tlz', '.tar.lz', '.tar.lzma')
+ZIP_EXTENSIONS = ('.zip', WHEEL_EXTENSION)
+TAR_EXTENSIONS = ('.tar.gz', '.tgz', '.tar')
+ARCHIVE_EXTENSIONS = (
+    ZIP_EXTENSIONS + BZ2_EXTENSIONS + TAR_EXTENSIONS + XZ_EXTENSIONS)
+SUPPORTED_EXTENSIONS = ZIP_EXTENSIONS + TAR_EXTENSIONS
+
+try:
+    import bz2  # noqa
+    SUPPORTED_EXTENSIONS += BZ2_EXTENSIONS
+except ImportError:
+    logger.debug('bz2 module is not available')
+
+try:
+    # Only for Python 3.3+
+    import lzma  # noqa
+    SUPPORTED_EXTENSIONS += XZ_EXTENSIONS
+except ImportError:
+    logger.debug('lzma module is not available')
+
+
+def unpack_file_url(
```

Comment:
It would be best if the functions in the module are ordered so that functions only depend on function that occur above them in the module, so ordered low-level to high-level. That way the module can be read top-to-bottom, and it becomes easier to locate things.

Also, the tests of the functions moved to this module should be moved to e.g. `test_unpacking.py`, following the pattern of other test modules.